instrumentation/rocketmq/rocketmq-client-java/rocketmq-client-java-5.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/rocketmqclientjava/v5_0/RocketMqProducerInstrumentation.java

```diff
@@ -0,0 +1,156 @@
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeInstrumentation;
+import io.opentelemetry.javaagent.extension.instrumentation.TypeTransformer;
+import java.util.ArrayList;
+import java.util.List;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.rocketmq.client.java.impl.producer.SendReceiptImpl;
+import org.apache.rocketmq.client.java.message.PublishingMessageImpl;
+import org.apache.rocketmq.shaded.com.google.common.util.concurrent.FutureCallback;
+import org.apache.rocketmq.shaded.com.google.common.util.concurrent.Futures;
+import org.apache.rocketmq.shaded.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.rocketmq.shaded.com.google.common.util.concurrent.SettableFuture;
+
+public class RocketMqProducerInstrumentation implements TypeInstrumentation {
+
+  /** Match the implementation of RocketMQ producer. */
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return named("org.apache.rocketmq.client.java.impl.producer.ProducerImpl");
+  }
+
+  @Override
+  public void transform(TypeTransformer transformer) {
+    transformer.applyAdviceToMethod(
+        isMethod()
+            .and(named("send0"))
+            .and(takesArguments(6))
+            .and(
+                takesArgument(
+                    0,
+                    named(
+                        "org.apache.rocketmq.shaded.com.google.common.util.concurrent.SettableFuture")))
+            .and(takesArgument(1, String.class))
+            .and(takesArgument(2, named("org.apache.rocketmq.client.java.message.MessageType")))
+            .and(takesArgument(3, List.class))
+            .and(takesArgument(4, List.class))
+            .and(takesArgument(5, int.class)),
+        RocketMqProducerInstrumentation.class.getName() + "$StartAdvice");
+  }
+
+  @SuppressWarnings("unused")
+  public static class StartAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void onEnter(
+        @Advice.Argument(0) SettableFuture<List<SendReceiptImpl>> future0,
+        @Advice.Argument(4) List<PublishingMessageImpl> messages) {
+      Instrumenter<PublishingMessageImpl, SendReceiptImpl> instrumenter =
+          RocketMqSingletons.producerInstrumenter();
+      int count = messages.size();
+      List<SettableFuture<SendReceiptImpl>> futures = FutureConverter.covert(future0, count);
+      for (int i = 0; i < count; i++) {
+        PublishingMessageImpl message = messages.get(i);
+
+        // Try to extract parent context from message.
+        Context parentContext = ParentContextExtractor.fromMessage(message);
```

Comment:
I don't follow why this would be needed. I'd expect `Context.current()` to be used as parent context.

Reply:
Especially, we allow user customize the trace parent of the message sending span rather than propagate from `Context.current`.

Reply:
Actually I am confused about that, the trace context of OpenTelemetry is propagated by `ThreadLocal` in default, but the thread to send message in the implementation of RocketMQ is the not the thread user invoked `Producer#send`, so how does the producer inherit the trace context from the upstream?

Reply:
this is tricky. you may have to introduce instrumentation on `Producer#send` to capture `Context.current()`, attach that to the Message (or some object that travels down to your existing instrumentation) and then in your existing instrumentation, pull the Context out to use it

Reply:
OK, I will implement it as you suggested with `VirtualField`.

Reply:
Done